### PR TITLE
feat: incorporate better-ajv-errors

### LINKED ISSUE #221: Clearer errors that pinpoint exact issue without extraneous errors
### **I'm submitting a...**
  - [ ] bug report
  - [x] feature request

### What is the current behavior?
When the schema is invalid, multiple seemingly-unrelated errors are returned along with the true error. These extra errors make a bit harder to pinpoint the actual issue.

### What is the expected behavior?
Only the actual error / root cause is listed as an error.

### What is the motivation / use case for changing the behavior?
Faster debugging of errors returned by Spectral. Fellow engineers would be able to more quickly fix issues in their schemas if the one actual error is provided.

### Please tell us about your environment:
```
$ spectral --version
@stoplight/spectral/3.0.0 linux-x64 node-v12.2.0
```

### Other information

Here's a sample schema:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Swagger Petstore
  description: Swagger Petstore
  license:
    name: MIT
  contact:
    name: Test
servers:
  - url: http://petstore.swagger.io/v1
paths:
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      description: Get a pet
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                type: array
                additionalItems: false
                items:
                  required: []
                  properties:
                    id:
                      type: integer
```

There are two issues with this schema:
1. `additionalItems` is not valid OpenAPI
2. `required` cannot be an empty list

Here is what spectral returns:
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 32:22  error  oas3-schema  should NOT have additional properties: additionalItems
 32:22  error  oas3-schema  should have required property '$ref'
 32:22  error  oas3-schema  should match exactly one schema in oneOf

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the first error is the only one that from my perspective is relevant and makes sense.

If that's fixed (remove `additionalItems`), then: the error is
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 34:23  error  oas3-schema  should have required property '$ref'
 34:23  error  oas3-schema  should match exactly one schema in oneOf
 35:28  error  oas3-schema  should NOT have fewer than 1 items

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the last listed error is the "real" error. As a first time user, this was confusing. The first two listed errors sent me chasing something that wasn't there.

Great project, btw!

### PR DESCRIPTION
**Pull request by P0lip**
### Please check if the PR fulfills these requirements
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

PR attempting to fix #221.

### What is the current behavior?

Spectral produces unuseful schema validation messages.

### If this is a feature change, what is the new behavior?

We integrate better-ajv-errors, which should, in theory, filter out the messages and provide more meaningful output.

### Does this PR introduce a breaking change?

It does not.

What changes might users need to make in their application due to this PR?

### Other information

(e.g. detailed explanation, related issues, links for us to have context, eg. stackoverflow, issues outside of the repo, forum, etc.)

### CODE REVIEW

**Comment by P0lip** [comment]
@philsturgeon what is your take on the new validation messages we generate?
I'm not quite satisfied with what better-ajv-errors generate, but it's just my humble opinion.

File: src/functions/schema.ts [hunk 1]
```diff
@@ -59,33 +55,16 @@ export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
...
-          message += `: ${(error.params as AJV.AdditionalPropertiesParams).additionalProperty}`;
-        }
-
-        collectedErrors.push(error.keyword);
-        results.push({
-          path: [...path, ...formatPath(error.dataPath)],
-          message,
-        });
-      }
+    const validator = validators.get(schemaObj);
```
Comment (philsturgeon) [inline src/functions/schema.ts]: Could we throw in some code comments that explain a tiny bit of "what" but mostly "why"? Both are a little unclear.

**Comment by philsturgeon** [comment]
@P0lip hey I'm gonna try and get a beta out shortly, can you get them comments in and we'll merge this? I don't want future people tripping over your advanced TS. :)

File: src/functions/schema.ts (hunk 1 above)
Comment (P0lip) [reply src/functions/schema.ts]: @philsturgeon sure, added. The main reason behind making this change is a very old todo comment we used to have here. The comment was about using `ajv.compile` and this is exactly what the following PR implements.

**Comment by P0lip** [comment]
@philsturgeon I'm going to add a few more integration tests. We already have a fairly alright coverage of schema function, but we are a bit lenient in integration tests assertion if you know what I mean.

**Comment by philsturgeon** [comment]
You broke it! :D

**Pull request merged**